mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala

```diff
@@ -281,7 +281,11 @@ class DenseMatrix(
   /**
    * @since 1.3.0
    */
-  override def apply(i: Int, j: Int): Double = values(index(i, j))
+  override def apply(i: Int, j: Int): Double = {
+    require(i < numRows && i >=0, s"Expected i to be non-negative and < $numRows, got $i")
```

Comment:
Actually, perhaps we should move this check into `index` so other methods (e.g. `update`) will also get checked